File: apps/bridge/src/standup.ts
// Standup and discuss commands — TypeScript implementation.
// /standup [@agent ...]   → 24h hive_mind status per agent; Main consolidates.
// /discuss <topic>        → each agent weighs in; Main consolidates.
//
// Agents run in parallel (no cross-talk until consolidation).

import type Database from "better-sqlite3";
import { query } from "@anthropic-ai/claude-agent-sdk";
import { resolve } from "node:path";

export const ALL_AGENTS = ["main", "comms", "content", "ops", "research"] as const;
export type AgentName = (typeof ALL_AGENTS)[number];

export function parseStandupCommand(
  text: string,
): { cmd: "standup" | "discuss"; agents: AgentName[]; topic: string } | null {
  const t = text.trim();
  if (!t.startsWith("/")) return null;

  const standupMatch = t.match(/^\/standup(?:\s+(.+))?$/i);
  if (standupMatch) {
    const rest = standupMatch[1] ?? "";
    const tagged = rest
      .split(/\s+/)
      .map((s) => s.replace(/^@/, "").toLowerCase())
      .filter((s): s is AgentName => s !== "main" && ALL_AGENTS.includes(s as AgentName));
    return {
      cmd: "standup",
      agents: tagged.length ? tagged : ALL_AGENTS.filter((a) => a !== "main"),
      topic: "",
    };
  }

  const discussMatch = t.match(/^\/discuss\s+(.+)$/i);
  if (discussMatch) {
    return {
      cmd: "discuss",
      agents: ALL_AGENTS.filter((a) => a !== "main"),
      topic: discussMatch[1].trim(),
    };
  }

  return null;
}

function fetch24h(
  db: Database.Database,
  agent: string,
): Array<{ prompt: string; reply: string }> {
  const cutoff = Math.floor(Date.now() / 1000) - 86400;
  return db
    .prepare(
      `SELECT prompt, reply FROM hive_mind
       WHERE agent = ? AND created_at >= ?
       ORDER BY created_at DESC LIMIT 50`,
    )
    .all(agent, cutoff) as Array<{ prompt: string; reply: string }>;
}

function standupPrompt(
  agent: string,
  entries: Array<{ prompt: string; reply: string }>,
): string {
  if (!entries.length) {
    return (
      `You are the ${agent} agent. You have nothing in hive_mind for the last 24h. ` +
      `Reply with exactly: 'Nothing to report.'`
    );
  }
  const bullets = entries
    .slice(0, 20)
    .map((e) => `- ${e.prompt.slice(0, 80)} → ${e.reply.slice(0, 120)}`)
    .join("\n");
  return (
    `You are the ${agent} agent. Give a 3-bullet standup of your last 24h. Terse, no preamble.\n` +
    `Format:\nWRAPPED: <one line>\nQUEUED: <one line>\nBLOCKED: <one line or 'none'>\n\n` +
    `Hive_mind activity:\n${bullets}`
  );
}

function discussPrompt(agent: string, topic: string): string {
  return (
    `You are the ${agent} agent. Topic: ${topic}\n\n` +
    `Give your unique angle in 2–3 sentences. Stay in your lane (${agent}'s expertise). ` +
    `Don't summarize others — you can't see them.`
  );
}

function consolidatePrompt(replies: Record<string, string>, topic: string): string {
  const formatted = Object.entries(replies)
    .map(([a, r]) => `## ${a}\n${r}`)
    .join("\n\n");
  if (topic) {
    return (
      `You are Main. Five agents weighed in on: ${topic}\n\n${formatted}\n\n` +
      `Consolidate in <=4 sentences. Surface the strongest disagreement. ` +
      `State your call. No preamble.`
    );
  }
  return (
    `You are Main. Standup replies:\n\n${formatted}\n\n` +
    `Synthesize: who's blocked, who needs attention, what should Jay do today. <=4 sentences.`
  );
}

async function runQuery(agentCwd: string, prompt: string): Promise<string> {
  let reply = "";
  const result = query({ prompt, options: { cwd: agentCwd } });
  for await (const ev of result) {
    if (ev.type === "assistant" && ev.message?.content) {
      for (const block of ev.message.content) {
        if (block.type === "text") reply += block.text;
      }
    }
  }
  return reply.trim() || "(no reply)";
}

export async function runStandupCommand(
  db: Database.Database,
  agentRoot: string,
  agents: AgentName[],
  topic: string,
  cmd: "standup" | "discuss",
): Promise<string> {
  // Run specialist agents in parallel — no cross-talk until Main consolidates.
  const replyEntries = await Promise.all(
    agents.map(async (agent) => {
      const prompt =
        cmd === "standup"
          ? standupPrompt(agent, fetch24h(db, agent))
          : discussPrompt(agent, topic);
      const reply = await runQuery(resolve(agentRoot, agent), prompt);
      return [agent, reply] as const;
    }),
  );

  const replies = Object.fromEntries(replyEntries);

  // Main consolidates all replies.
  const mainPrompt = consolidatePrompt(replies, topic);
  const mainReply = await runQuery(resolve(agentRoot, "main"), mainPrompt);

  const sections = replyEntries.map(([agent, reply]) => `*${agent}:* ${reply}`).join("\n\n");
  return `${sections}\n\n---\n*Main:* ${mainReply}`;
}

File: apps/warroom/standup.py
"""
War Room slash commands — /standup and /discuss.

V3 transcript ch.3, ch.16:
- /standup [@agent ...]   → each agent gives a 24h status report from hive_mind.
                            Main consolidates if no specific agents tagged.
- /discuss <topic>        → each agent weighs in with their angle. Main consolidates last.

Both commands run agents in isolation (no cross-talk during their reply),
then route through Main for final consolidation. The agent doesn't see
other replies until consolidation, mirroring the standup discipline of
"give your update, don't react."
"""

import os
import re
import sqlite3
import time
from typing import Optional

DB_PATH = os.environ.get("DB_PATH", "./store/claudeclaw.db")
ALL_AGENTS = ["main", "comms", "content", "ops", "research", "meta"]


def parse_command(text: str) -> Optional[dict]:
    """Returns {'cmd': 'standup'|'discuss', 'agents': [...], 'topic': str} or None."""
    t = text.strip()
    if not t.startswith("/"):
        return None

    standup_match = re.match(r"^/standup(?:\s+(.+))?$", t, re.IGNORECASE)
    if standup_match:
        rest = standup_match.group(1) or ""
        # Accept space- or comma-separated tags, with or without leading @.
        tokens = re.split(r"[\s,]+", rest)
        agents = [
            tok.lstrip("@").lower()
            for tok in tokens
            if tok.lstrip("@").lower() in ALL_AGENTS
        ]
        return {
            "cmd": "standup",
            "agents": agents or [a for a in ALL_AGENTS if a != "main"],
            "topic": "",
        }

    discuss_match = re.match(r"^/discuss\s+(.+)$", t, re.IGNORECASE)
    if discuss_match:
        return {
            "cmd": "discuss",
            "agents": [a for a in ALL_AGENTS if a != "main"],
            "topic": discuss_match.group(1).strip(),
        }

    return None


def fetch_24h_summary(agent: str) -> list[dict]:
    """Pull the last 24h of hive_mind entries for one agent."""
    conn = sqlite3.connect(DB_PATH)
    conn.row_factory = sqlite3.Row
    cutoff = int(time.time()) - 86400
    rows = conn.execute(
        """
        SELECT id, prompt, reply, created_at
        FROM hive_mind
        WHERE agent = ? AND created_at >= ?
        ORDER BY created_at DESC
        LIMIT 50
        """,
        (agent, cutoff),
    ).fetchall()
    conn.close()
    return [dict(r) for r in rows]


def standup_prompt(agent: str, entries: list[dict]) -> str:
    """Per-agent standup prompt. Each agent runs in isolation."""
    if not entries:
        return (
            f"You are the {agent} agent. You have nothing in hive_mind for the last 24h. "
            f"Reply with one sentence: 'Nothing to report.'"
        )
    bullets = "\n".join(
        f"- {e['prompt'][:80]} → {e['reply'][:120]}" for e in entries[:20]
    )
    return (
        f"You are the {agent} agent. Give a 3-bullet standup of what you've done in the last 24h. "
        f"Be terse, no fluff, no preamble. Format:\n"
        f"WRAPPED: <one line>\nQUEUED: <one line>\nBLOCKED: <one line or 'none'>\n\n"
        f"Hive_mind activity:\n{bullets}"
    )


def discuss_prompt(agent: str, topic: str) -> str:
    return (
        f"You are the {agent} agent. The topic: {topic}\n\n"
        f"Give your unique angle in 2–3 sentences. Stay in your lane "
        f"({agent}'s expertise). Don't summarize others; you can't see them."
    )


def consolidate_prompt(replies: dict[str, str], topic: str = "") -> str:
    """Main agent's consolidation prompt. Sees all replies."""
    formatted = "\n\n".join(f"## {a}\n{r}" for a, r in replies.items())
    if topic:
        return (
            f"You are Main. Five agents weighed in on: {topic}\n\n"
            f"{formatted}\n\n"
            f"Consolidate in <=4 sentences. Surface the strongest disagreement. "
            f"Then state your call. No preamble."
        )
    return (
        f"You are Main. Standup replies:\n\n{formatted}\n\n"
        f"Synthesize: who's blocked, who needs attention, what should Jay do today. <=4 sentences."
    )
